sklearn/cluster/_k_means_fast.pxd

```diff
@@ -0,0 +1,23 @@
+# cython: language_level=3
+
+
+from cython cimport floating
+cimport numpy as np
+
+
+cdef floating _euclidean_dense_dense(floating*, floating*, int, bint) nogil
+
+cdef floating _euclidean_sparse_dense(floating[::1], int[::1], floating[::1],
+                                      floating, bint) nogil
+
+cpdef void _relocate_empty_clusters_dense(
+    np.ndarray[floating, ndim=2, mode='c'], floating[::1], floating[:, ::1],
```

Comment:
It would fail if X is read only. It's still the same issue of const fused type memoryviews

Reply:
aah, right (wasn't that fixed?? I don't remember the status)

Maybe add a small comment at the top of the .pxd or the pyx to say why X is an array rather than a view. maybe there's already one that I mised

Reply:
fixed in cython master but not released. Will be in cython 0.30. will be in sklearn when our min cython version is 0.3